### ISSUE TITLE
opengl problem when launching "clickable --desktop"

### ISSUE THREAD

**Issue by luksus42** [bhdouglass/clickable#44]
Hi, 
I have the problem, that I get opengl errors, when trying to launch a click app via "clickable --desktop":
`libGL error: No matching fbConfigs or visuals found`
`libGL error: failed to load driver: swrast`
`Unrecognized OpenGL version`
`Unrecognized OpenGL version`

So I discovered, it has something to do with nvidia-drivers and the docker container needs the same driver like the host.
Therefore I installed the package "nvidia-docker" and recreated the container-image with options:
`# nvidia-container-runtime`
`ENV NVIDIA_VISIBLE_DEVICES all`
`ENV NVIDIA_DRIVER_CAPABILITIES compute,utility`

Now I can start the container and watch "glxgears" with:
`docker run --runtime=nvidia -e NVIDIA_VISIBLE_DEVICES=all --rm clickable/ubuntu-sdk:15.04-amd64 apt install mesa-utils && glxgears`

**But** I recently discovered that this even works without that nvidia specific commands... ;-)

So my first intention was to ask if there is a way to serve that nvidia commands into the clickable application.
Since this doesn't seem to be necessary, I'm hoping you got any idea, what could solve my issue.

Thank you very much.
(I'm on manjaro linux)

**Comment by luksus42** [comment]
ahh sorry... just saw the same issue as "closed"...
I close this.

**Comment by bhdouglass** [comment]
I don't have a problem adding this stuff to clickable, I just don't have a way to test it. But it looks like you've done most of the leg work.

**Comment by luksus42** [comment]
I could test it.
As I understand it's needed either to create the container with that environment-arguments, or to start docker with that --runtime parameters.
So it would be useful to be able to serve some docker related extra arguments when launching clickable.

But anyway, this doesn't seem to fix the initial opengl issue...

**Comment by bhdouglass** [comment]
So running docker with `--runtime=nvidia` is all I need to add to fix the issue?

**Comment by luksus42** [comment]
No. That would just be the way to start the container with nvidia-drivers (if nvidia-docker is installed). 
But as I already wrote: that did not fix the opengl issue for me.

Perhaps I am able to evaluate how it would work the right way.
Could you please write down the docker command how I execute docker while serving a click-package to it?

**Comment by bhdouglass** [comment]
Ok, so I guess the first question is, can you get glxgears running on a regular container (maybe that glvnd one?). Then maybe we can see if it can be ported over to the clickable container.

**Comment by luksus42** [comment]
Yes, that is the question.
If I get it running, I will report back here.

**Comment by bhdouglass** [comment]
Good luck! And thanks for looking into this!

**Comment by luksus42** [comment]
I got it.

Tested with nvidia-docker version 1.0.1:

**First, I needed to create a new container, with opengl and nvidia settings included:**
[Dockerfile.txt](https://github.com/bhdouglass/clickable/files/1864395/Dockerfile.txt)
(nvidia-part is taken from here: https://github.com/NVIDIA/nvidia-docker/blob/0158f50a75bef75bb96c03ebe4c29fb7e9a2ca13/ubuntu-14.04/opengl/runtime/Dockerfile)

Since there are no special drivers built in, I think the resulting container should be universally applicable for the clickable package.

**Then it's necessary to start the container with `nvidia-docker` command instead of `docker`.**

The arguments-list is very similar to the one that "clickable" creates:
```bash
nvidia-docker run \
-e DISPLAY=:0 \
-v /home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog:/home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog \
-v /tmp/.X11-unix:/tmp/.X11-unix \
-v /tmp/.docker.xauth:/tmp/.docker.xauth \
-v /tmp/clickable/share:/tmp/.local/share \
-v /tmp/clickable/cacge:/tmp/.cache \
-v /tmp/clickable/config:/tmp/.config \
-v /etc/passwd:/etc/passwd \
-e XAUTHORITY=/tmp/.docker.xauth \
-e QML2_IMPORT_PATH=/home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog/build/tmp/lib/x86_64-linux-gnu \
-e LD_LIBRARY_PATH=/home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog/build/tmp/lib/x86_64-linux-gnu:/usr/local/nvidia/lib:/usr/local/nvidia/lib64 \
-e PATH=/usr/local/nvidia/bin:/bin:/usr/bin:/home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog/build/tmp/bin:/home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog/build/tmp/lib/x86_64-linux-gnu/bin:/home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog/build/tmp \
-e HOME=/tmp \
-e OXIDE_NO_SANDBOX=1 \
-w /home/luke/Daten/EigeneDateien/Code/Ubuntu/SoundRecog/soundrecog/build/tmp \
-u 1000 \
--rm \
-i clickable/ubuntu-sdk:15.04-amd64 bash \
-c "webapp-container --inspector --app-id=dummy ./www/index.html"
```
The only changes are:
- `/usr/local/nvidia/bin` added to **PATH**
- `/usr/local/nvidia/lib:/usr/local/nvidia/lib64`  added to **LD_LIBRARY_PATH** 
- the **--app-id** argument for the webapp_container was needed to get the app running (or to finally see the app, I'm not shure what happened in the background)

Then I modified the clickable-script, that the arguments are being generated that way.
So I am now able to execute `clickable --desktop` in my project-path's and get the apps running. :-)

I hope that helps, to implement some kind of "switch-to-nvidia-parameter"  into clickable...

**Comment by bhdouglass** [comment]
Very nice! Do you mind making a pull request with what you have? Then I'll see about getting it to play nicely with the existing setup.

**Comment by bhdouglass** [comment]
The new `--nvidia` option is now available in v4.1.0 (in the ppa soon!)
